idaes/core/util/model_diagnostics.py

```diff
@@ -4122,3 +4264,269 @@ def _collect_model_statistics(model):
     )
 
     return stats
+
+
+class ConstraintTermAnalysisVisitor(EXPR.StreamBasedExpressionVisitor):
+    """
+    Expression walker for checking Constraints for problematic terms.
+
+    This walker will walk the expression and look for summation terms
+    with mismatched magnitudes or potential cancellations.
+
+    Args:
+        term_mismatch_tolerance - tolerance to use when determining mismatched
+            terms
+        term_cancellation_tolerance - tolerance to use when identifying
+            possible cancellation of terms
+
+    Returns:
+        list of values for top-level summation terms
+        list of terms with mismatched magnitudes
+        list of terms with potential cancellations
+        bool indicating whether expression is a constant
+    """
+
+    def __init__(
+        self,
+        term_mismatch_tolerance: float = 1e6,
+        term_cancellation_tolerance: float = 1e-4,
+    ):
+        super().__init__()
+
+        self._mm_tol = log10(term_mismatch_tolerance)
+        self._sum_tol = term_cancellation_tolerance
+
+    def _check_base_type(self, node):
+        # [value], [mismatched terms], [canceling terms], constant
+        if isinstance(node, VarData):
+            const = node.fixed
+        else:
+            const = True
+        return [value(node)], [], [], const
+
+    def _get_value_for_sum_subexpression(self, child_data):
+        return sum(i for i in child_data[0])
+
+    def _check_sum_expression(self, node, child_data):
+        # Sum expressions need special handling
+        # For sums, collect all child values into a list
+        vals = []
+        mismatch = []
+        # We will check for cancellation in this node at the next level
+        # Pyomo is generally good at simplifying compound sums
+        cancelling = []
+        const = True
+        # Collect data from child nodes
+        for d in child_data:
+            vals.append(self._get_value_for_sum_subexpression(d))
+            mismatch += d[1]
+            cancelling += d[2]
+
+            # Expression is not constant if any child is not constant
+            if not d[3]:
+                const = False
+
+        # Check for mismatched terms
+        if len(vals) > 1:
+            absvals = [abs(v) for v in vals]
+            vl = max(absvals)
+            vs = min(absvals)
+            if vl != vs:
+                if vs == 0:
+                    diff = log10(vl)
+                else:
+                    diff = log10(vl / vs)
+
+                if diff >= self._mm_tol:
+                    mismatch.append(str(node))
+
+        return vals, mismatch, cancelling, const
+
+    def _check_product(self, node, child_data):
+        mismatch, cancelling, const = self._perform_checks(node, child_data)
+
+        val = self._get_value_for_sum_subexpression(
+            child_data[0]
+        ) * self._get_value_for_sum_subexpression(child_data[1])
+
+        return [val], mismatch, cancelling, const
```

Comment:
Almost all of your nonlinear handlers could be replaced by a single callback:
```python
def _check_general_expr(self, node, child_data):
    mismatch, cancelling, const = self._perform_checks(node, child_data)
    val = node._apply_operation(list(map(self._get_value_for_sum_subexpression, child_data)))
    return [val], mismatch, cancelling, const
```

Reply:
Thank you - I did not know I could do that.